apps/sync/js/adapters/bookmarks.js

```diff
@@ -0,0 +1,325 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/**
+  Bookmarks Data Adapter
+
+    The feature of Bookmarks adapter is to retrieve Bookmarks records from
+  FxSync Service and merge to Bookmarks DataStore ('firefox-sync-bookmarks')
+  for FxSync only which has the same functionality, and the adapter is based on
+  SyncEngine's design to handle 'update' and 'handleConflict' request. Bookmarks
+  Data Adapter provides Read-Only version currently, and the basic ideas are
+  included:
+  - Traverse all records from Bookmarks Collection (BC).
+  - Add a new or existed bookmark to Bookmarks DataStore (BDS).
+    - Convert the records from BC to BDS format.
+    - Merge a new record and the existed one then write to BDS.
+  - Remove an existed bookmark when receiving a "deleted: true" record.
+**/
+
+'use strict';
+
+/* global
+  asyncStorage,
+  DataAdapters,
+  LazyLoader
+*/
+
+const BOOKMARKS_COLLECTION_MTIME = 'collections::bookmarks::mtime';
+const BOOKMARKS_SYNCTOID_PREFIX = 'SynctoId::bookmarks::';
+
+var BookmarksHelper = (() => {
+  var bookmarksStore;
+  function _ensureStore() {
+    if (bookmarksStore) {
+      return Promise.resolve(bookmarksStore);
+    }
+    return navigator.getDataStores('firefox-sync-bookmarks').then(stores => {
+      bookmarksStore = stores[0];
```

Comment:
Could you check that we are the owners of this DataStore, please?

This is a privacy issue that is quite unlikely to happen, but just in case DataStore is ever opened to non certified apps we should be ready for it. If another app creates a DataStore with the same name, we might end up writing all the user's bookmarks in that DataStore.